source/lac/matrix_scaling.cc

```diff
@@ -0,0 +1,1373 @@
+  else if (norm_type == "linfty")
+    return row_col_norm.linfty_norm() < control.scaling_tolerance;
+  else
+    {
+      Assert(false, ExcNotImplemented());
+      return false;
+    }
+}
+
+
+
+template <typename Number>
+bool
+MatrixScaling::check_convergence(Vector<Number>     row_norm,
+                                 Vector<Number>     col_norm,
+                                 const std::string &norm_type) const
+{
+  row_norm.add(-1.0);
+  col_norm.add(-1.0);
+  if (norm_type == "l1")
+    return (row_norm.l1_norm() < control.scaling_tolerance &&
+            col_norm.l1_norm() < control.scaling_tolerance);
+  else if (norm_type == "linfty")
+    return (row_norm.linfty_norm() < control.scaling_tolerance &&
+            col_norm.linfty_norm() < control.scaling_tolerance);
+  else
+    {
+      Assert(false, ExcNotImplemented());
+      return false;
+    }
+}
+
+
+
+bool
+MatrixScaling::check_convergence(Vector<double>     local_row_col_norm,
+                                 const std::string &norm_type,
+                                 const MPI_Comm     mpi_communicator) const
+{
+  local_row_col_norm.add(-1.0);
+  bool local_not_converged;
+
+  if (norm_type == "l1")
+    local_not_converged =
+      !(local_row_col_norm.l1_norm() < control.scaling_tolerance);
+  else if (norm_type == "linfty")
+    local_not_converged =
+      !(local_row_col_norm.linfty_norm() < control.scaling_tolerance);
+  else
+    {
+      Assert(false, ExcNotImplemented());
+      return false;
+    }
+
+  bool any_not_converged =
+    Utilities::MPI::logical_or(local_not_converged, mpi_communicator);
+
+  return !any_not_converged;
+}
+
+
+
+bool
+MatrixScaling::check_convergence(Vector<double>     local_row_norm,
+                                 Vector<double>     local_col_norm,
+                                 const std::string &norm_type,
+                                 const MPI_Comm     mpi_communicator) const
+{
+  local_row_norm.add(-1.0);
+  local_col_norm.add(-1.0);
+  bool local_not_converged;
+
+  if (norm_type == "l1")
+    local_not_converged =
+      !(local_row_norm.l1_norm() < control.scaling_tolerance &&
+        local_col_norm.l1_norm() < control.scaling_tolerance);
+  else if (norm_type == "linfty")
+    local_not_converged =
+      !(local_row_norm.linfty_norm() < control.scaling_tolerance &&
+        local_col_norm.linfty_norm() < control.scaling_tolerance);
+  else
+    {
+      Assert(false, ExcNotImplemented());
+      return false;
+    }
+
+  bool any_not_converged =
+    Utilities::MPI::logical_or(local_not_converged, mpi_communicator);
+
+  return !any_not_converged;
+}
+
+
+
+void
+MatrixScaling::send_prepare_col_norms(
+  std::map<unsigned int,
+           std::vector<std::pair<types::global_dof_index, double>>> &send_data,
+  const std::map<types::global_dof_index, double> &partial_column_norms,
+  Vector<double>                                  &local_col_norms)
```

Comment:
I think that the first and third argument are output arguments, and the second is an input argument, right? If so, sort them so that the input arguments come first, and the output arguments last.